source/common/network/utility.cc

```diff
@@ -208,6 +208,21 @@ Address::InstanceConstSharedPtr Utility::getLocalAddress(const Address::IpVersio
   return ret;
 }
 
+bool Utility::isLocalConnection(const Address::Instance& local_address,
+                                const Address::Instance& remote_address) {
```

Comment:
@PiotrSikora  You are correct that the name of the method is not reflecting what it actually does.
The closest to the truth would be to rename the config option to `source_destination_match` with possible enum values `{SAME, DISTINCT, ANY}` which would cover a lot of the use cases and be easily implemented in platform independent way

> You should really check if the IP address from socket.remoteAddress() matches IP address of one of local network interfaces.

This cannot be implemented in platform independent way. I found how it can be implemented on linux:
https://stackoverflow.com/questions/1160963/how-to-enumerate-all-ip-addresses-attached-to-a-machine-in-posix-c/1161018#1161018

There is a sample code which would need to be extended to support IPv6 if we want to go that way. This option can really be called isLocalConnecton. I am not sure what are the performance characteristics of this `ioctl` call.

I am fine with both solutions as well as having both of them together.

Thoughts?

Reply:
> The closest to the truth would be to rename the config option to source_destination_match with possible enum values {SAME, DISTINCT, ANY} which would cover a lot of the use cases and be easily implemented in platform independent way

But then you're limiting this feature by implementation and not by intent.

Most importantly, the current implementation doesn't work with `iptables` interception, which is a pretty big deal-breaker, IMHO.

> This cannot be implemented in platform independent way. I found how it can be implemented on linux:
> https://stackoverflow.com/questions/1160963/how-to-enumerate-all-ip-addresses-attached-to-a-machine-in-posix-c/1161018#1161018

That response is from 2009.

This can be easily done in a semi-portable way using `getifaddrs`, which we already use in Envoy (see: [Network::Utility::getLocalAddress](https://github.com/envoyproxy/envoy/blob/818ba8689fda0c0e0c1429969dc41fc42ca20849/source/common/network/utility.cc#L170)).

Reply:
@PiotrSikora I agree that it is better not to limit by implementation and have  the more general solution. 

Lets switch to the implementation as in `Network::Utility::getLocalAddress`
Then we can rename the configuration parameter to `source_type : {LOCAL, EXTERNAL, ANY}`

Do we have any concerns about the performance of calling `getifaddrs` on every accepted connection?

Reply:
I don't know the cost of `getifaddrs()`, so it's hard to tell, but we could always cache the results for X seconds if it turns out to be expensive.

In any case, we should only call it if either `LOCAL` or `EXTERNAL` are configured, and provide fast-path that doesn't do the check if only `ANY` (default) is configured.

Reply:
@PiotrSikora  

I agree with the potential optimization but for our use cаse we will always be in the bad case.

There is one further optimization which is when `src == dst` we know it has to be local so we can do the interface search only when `src != dst` and `LOCAL` or `EXTERNAL` is configured.

I am strongly against caching this information because it would lead to random failures which are very hard to debug. I would rather have both solutions in and clearly explain the difference in functionality and performance. One is `source` to `destination` matching and the other is matching source against the list of local interfaces.

I am fine though with implementing the interface matching first and we can benchmark the performance impact of it. Then we can decide if there is a need for the `source == destination mаtching option.